File: Sources/CodexTopCore/RolloutReducer.swift
import Foundation
import Darwin

/// Consumes only event metadata. Message bodies and tool arguments are never retained.
public struct RolloutReducer: Sendable {
    public private(set) var activity = TaskActivity()
    public private(set) var quota: QuotaSnapshot?
    private var waitingCallIDs: Set<String> = []
    private var asynchronousQuestion = false
    public init() {}

    public mutating func consume(_ line: Data) {
        guard let value = try? JSONSerialization.jsonObject(with: line) as? [String: Any],
              let payload = value["payload"] as? [String: Any],
              let kind = value["type"] as? String else { return }
        let at = Self.date(value["timestamp"])
        // Old records may arrive after newer records during a compaction/replay.
        if let at, let last = activity.lastEventAt, at < last { return }
        let type = payload["type"] as? String ?? ""
        if kind == "event_msg", ["task_complete", "turn_complete", "turn_aborted", "task_cancelled", "turn_cancelled", "task_failed", "turn_failed"].contains(type),
           let turn = payload["turn_id"] as? String, let current = activity.turnID, turn != current { return }
        let userItem = type == "item_completed" && ((payload["item"] as? [String: Any])?["type"] as? String)?.lowercased() == "usermessage"
        let responseUserMessage = kind == "response_item" && type == "message" && (payload["role"] as? String) == "user"
        let explicitContinuation = responseUserMessage || (kind == "event_msg" && (["task_started", "turn_started", "user_message", "user_input", "approval_resolved"].contains(type) || userItem))
        // A trailing explanation or tool result does not prove a failed turn resumed.
        if activity.phase == .failed && !explicitContinuation {
            if kind == "event_msg", type == "token_count", let limits = payload["rate_limits"] as? [String: Any], let at {
                updateQuota(limits, at: at); activity.lastEventAt = at
            }
            return
        }
        if kind == "event_msg" {
            switch type {
            case "task_started", "turn_started":
                activity = TaskActivity(phase: .running, detail: "正在执行", lastEventAt: at, startedAt: at)
                activity.turnID = payload["turn_id"] as? String
                waitingCallIDs.removeAll()
                asynchronousQuestion = false
            case "task_complete", "turn_complete":
                if asynchronousQuestion { waiting(at, detail: "等待你的回答") }
                else {
                    activity.phase = .completed; activity.detail = "本轮执行已结束"
                    activity.waitingStartedAt = nil
                }
                waitingCallIDs.removeAll()
            case "turn_aborted", "task_cancelled", "turn_cancelled":
                activity.phase = .stopped; activity.detail = "本轮执行已停止"; waitingCallIDs.removeAll()
                activity.waitingStartedAt = nil; asynchronousQuestion = false
            case "task_failed", "turn_failed":
                activity.phase = .failed; activity.detail = "执行遇到问题，请回到 Codex 查看"; waitingCallIDs.removeAll()
                activity.waitingStartedAt = nil; asynchronousQuestion = false
            case "request_user_input", "user_input_requested", "exec_approval_request", "apply_patch_approval_request":
                waiting(at, detail: "等待你的输入或确认")
            case "user_message", "user_input", "approval_resolved":
                active(at, detail: "收到输入，正在继续"); waitingCallIDs.removeAll(); asynchronousQuestion = false
            case "agent_message", "agent_reasoning":
                if activity.phase != .waiting && !activity.phase.isFinished { active(at, detail: "正在处理任务") }
            case "token_count":
                if let limits = payload["rate_limits"] as? [String: Any], let at { updateQuota(limits, at: at) }
            case "item_completed":
                if let item = payload["item"] as? [String: Any] {
                    let itemType = (item["type"] as? String ?? "").lowercased()
                    if itemType == "usermessage" { active(at, detail: "收到输入，正在继续"); waitingCallIDs.removeAll(); asynchronousQuestion = false }
                    else if ["commandexecution", "filechange", "reasoning", "mcptoolcall"].contains(itemType), activity.phase != .waiting, !activity.phase.isFinished {
                        active(at, detail: itemType == "filechange" ? "正在修改文件" : "正在执行任务")
                    }
                }
            default: return
            }
            if let at { activity.lastEventAt = at }
        } else if kind == "response_item" {
            if responseUserMessage {
                // The response record can precede the corresponding UserMessage event.
                active(at, detail: "收到输入，正在继续")
                waitingCallIDs.removeAll(); asynchronousQuestion = false
                if let at { activity.lastEventAt = at }
            } else if type == "function_call" || type == "custom_tool_call" {
                let name = payload["name"] as? String ?? ""
                if name == "request_user_input" || name == "request_user_input_async" || name.hasSuffix("__request_user_input") {
                    waiting(at, detail: "等待你的回答")
                    if name == "request_user_input_async" { asynchronousQuestion = true }
                    else if let id = payload["call_id"] as? String { waitingCallIDs.insert(id) }
                } else if activity.phase != .waiting { active(at, detail: "正在执行任务") }
                if let at { activity.lastEventAt = at }
            } else if type == "function_call_output" || type == "custom_tool_call_output" {
                if let id = payload["call_id"] as? String, waitingCallIDs.remove(id) != nil, waitingCallIDs.isEmpty, !asynchronousQuestion {
                    active(at, detail: "已收到回答，正在继续")
                }
                if let at { activity.lastEventAt = at }
            }
        }
    }
    private mutating func active(_ at: Date?, detail: String) {
        if activity.phase.isFinished || activity.phase == .failed { activity.startedAt = at; activity.turnID = nil }
        activity.phase = .running; activity.detail = detail
        activity.waitingStartedAt = nil
    }
    private mutating func waiting(_ at: Date?, detail: String) {
        if activity.phase != .waiting {
            // A terminal turn cannot supply a start for a new wait with no observed start event.
            if activity.phase.isFinished { activity.startedAt = nil; activity.turnID = nil }
            activity.waitingStartedAt = at
        }
        // A missing first timestamp stays unknown; a repeated wait must not invent a later anchor.
        activity.phase = .waiting; activity.detail = detail
    }
    fileprivate mutating func recoverTiming(from recovered: TaskActivity) {
        guard activity.startedAt == nil, activity.phase == .running || activity.phase == .waiting,
              recovered.phase == activity.phase, let last = activity.lastEventAt,
              recovered.lastEventAt == last, let start = recovered.startedAt,
              start.timeIntervalSinceReferenceDate.isFinite, start <= last,
              activity.turnID == nil || recovered.turnID == activity.turnID,
              activity.waitingStartedAt.map({ start <= $0 }) ?? true else { return }
        activity.startedAt = start
        if activity.turnID == nil { activity.turnID = recovered.turnID }
    }
    private mutating func updateQuota(_ limits: [String: Any], at: Date) {
        if let id = limits["limit_id"] as? String, id != "codex" { return }
        var windows: [QuotaWindow] = []
        for key in ["primary", "secondary"] {
            guard let w = limits[key] as? [String: Any], let used = w["used_percent"] as? Double, used.isFinite else { continue }
            let minutes = (w["window_minutes"] as? Int) ?? ((w["limit_window_seconds"] as? Int).map { $0 / 60 })
            guard let minutes, minutes > 0 else { continue }
            windows.append(QuotaWindow(minutes: minutes, usedPercent: used, resetsAt: (w["resets_at"] as? Double).map(Date.init(timeIntervalSince1970:))))
        }
        if !windows.isEmpty { quota = QuotaSnapshot(observedAt: at, windows: windows) }
    }
    private static func date(_ value: Any?) -> Date? {
        guard let text = value as? String else { return nil }
        return (try? Date.ISO8601FormatStyle(includingFractionalSeconds: true).parse(text))
            ?? (try? Date.ISO8601FormatStyle().parse(text))
    }
}

public struct IncrementalRollout: Sendable {
    public private(set) var reducer = RolloutReducer()
    public private(set) var offset: UInt64 = 0
    /// The cursor reached the file size sampled by the latest refresh; a partial final line still waits for its newline.
    public private(set) var isCaughtUp = false
    private var inode: UInt64?
    private var observedSize: UInt64?
    private var modified: Date?
    private var pending = Data()
    private var skippingLongLine = false
    private var timingRecovery: RolloutTimingRecovery?
    private let maximumRead: Int
    private let maximumCatchUpRead: Int
    public init(maximumRead: Int = 65_536, maximumCatchUpRead: Int = 4 * 1_024 * 1_024) {
        self.maximumRead = max(1024, maximumRead)
        self.maximumCatchUpRead = max(self.maximumRead, maximumCatchUpRead)
    }
    /// Returns bytes actually read; unchanged files incur only a metadata check.
    public mutating func refresh(url: URL) throws -> Int {
        let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
        let size = (attributes[.size] as? NSNumber)?.uint64Value ?? 0
        let newInode = (attributes[.systemFileNumber] as? NSNumber)?.uint64Value
        let newModified = attributes[.modificationDate] as? Date
        if inode == newInode && size == offset && modified == newModified { return 0 }
        let rewritten = observedSize.map { size < $0 || (size == $0 && modified != newModified) } ?? false
        let reset = inode != newInode || size < offset || rewritten
        if reset { reducer = RolloutReducer(); offset = 0; pending.removeAll(); skippingLongLine = false; isCaughtUp = false; timingRecovery = nil }
        let handle = try FileHandle(forReadingFrom: url); defer { try? handle.close() }
        // Cold starts remain bounded. Warm reads must not skip events or discard pending questions/turn IDs.
        let start = reset && size > UInt64(maximumRead) ? size - UInt64(maximumRead) : offset
        let skipped = start > offset
        if skipped { pending.removeAll(); skippingLongLine = true }
        try handle.seek(toOffset: start)
        offset = start; inode = newInode; observedSize = size; modified = newModified
        let budget = reset ? maximumRead : maximumCatchUpRead
        var bytesRead = 0
        while offset < size && bytesRead < budget {
            let count = min(maximumRead, budget - bytesRead, Int(min(UInt64(maximumRead), size - offset)))
            let data = try handle.read(upToCount: count) ?? Data()
            guard !data.isEmpty else { break }
            offset += UInt64(data.count); bytesRead += data.count
            pending.append(data)
            while let newline = pending.firstIndex(of: 10) {
                let line = pending[..<newline]
                if !skippingLongLine && !line.isEmpty && line.count <= maximumRead { reducer.consume(Data(line)) }
                skippingLongLine = false
                pending.removeSubrange(...newline)
            }
            if pending.count > maximumRead { pending.removeAll(); skippingLongLine = true }
        }
        isCaughtUp = offset >= size
        if reducer.activity.startedAt != nil { timingRecovery = nil }
        return bytesRead
    }

    /// Optional work for selected tasks only. Progress and a failed search are
    /// retained across unchanged/appended files; replacement or rewrite resets them.
    public mutating func recoverTiming(url: URL, maximumBytes: Int = 4 * 1_024 * 1_024) throws -> Int {
        guard reducer.activity.startedAt == nil else { timingRecovery = nil; return 0 }
        guard isCaughtUp, reducer.activity.phase == .running || reducer.activity.phase == .waiting,
              timingRecovery?.isFinished != true else { return 0 }
        let attributes = try FileManager.default.attributesOfItem(atPath: url.path)
        let size = (attributes[.size] as? NSNumber)?.uint64Value ?? 0
        let currentInode = (attributes[.systemFileNumber] as? NSNumber)?.uint64Value
        let currentModified = attributes[.modificationDate] as? Date
        guard currentInode == inode, size >= offset,
              !(size == observedSize && currentModified != modified) else { return 0 }
        let handle = try FileHandle(forReadingFrom: url); defer { try? handle.close() }
        var before = stat()
        guard fstat(handle.fileDescriptor, &before) == 0,
              UInt64(before.st_ino) == inode, before.st_size >= 0,
              UInt64(before.st_size) >= offset else { return 0 }
        var recovery = timingRecovery ?? RolloutTimingRecovery(end: offset)
        let bytes = recovery.advance(handle: handle, through: offset, blockSize: maximumRead, budget: max(1_024, maximumBytes))
        var after = stat()
        let currentPath = try? FileManager.default.attributesOfItem(atPath: url.path)
        let pathInode = (currentPath?[.systemFileNumber] as? NSNumber)?.uint64Value
        let pathSize = (currentPath?[.size] as? NSNumber)?.uint64Value
        let pathModified = currentPath?[.modificationDate] as? Date
        guard fstat(handle.fileDescriptor, &after) == 0,
              before.st_ino == after.st_ino, after.st_size >= before.st_size,
              !(after.st_size == before.st_size && (after.st_mtimespec.tv_sec != before.st_mtimespec.tv_sec || after.st_mtimespec.tv_nsec != before.st_mtimespec.tv_nsec)),
              pathInode == inode, let pathSize, pathSize >= offset,
              !(pathSize == size && pathModified != currentModified) else {
            timingRecovery = nil
            return bytes
        }
        timingRecovery = recovery
        if let recovered = recovery.recovered { reducer.recoverTiming(from: recovered) }
        return bytes
    }
}

File: windows/src/CodexTop.Core/RolloutReducer.cs
using System.Text.Json;

namespace CodexTop.Core;

// State rules adapted from BuTangTang/Codex-Top (GPL-3.0).
public sealed class RolloutReducer
{
    public Activity Activity { get; private set; } = new();
    private readonly HashSet<string> synchronous = [];
    private readonly HashSet<string> asynchronous = [];
    private bool anonymousWait;
    public void Consume(ReadOnlySpan<byte> line)
    {
        JsonDocument doc;
        try { doc = JsonDocument.Parse(line.ToArray()); } catch (JsonException) { return; }
        using (doc)
        {
            var root = doc.RootElement; var payload = root.Field("payload");
            string? kind = root.Text("type"), type = payload.Text("type"), turn = payload.Text("turn_id");
            var at = root.Date("timestamp");
            if (at is { } time && Activity.LastEventAt is { } previous && time < previous) return;
            bool start = kind == "event_msg" && type is "task_started" or "turn_started";
            bool terminal = kind == "event_msg" && type is "task_complete" or "task_completed" or "turn_complete" or "turn_completed" or "turn_aborted" or "task_cancelled" or "turn_cancelled" or "task_failed" or "turn_failed";
            bool user = kind == "response_item" && type == "message" && payload.Text("role") == "user";
            if (!start && turn is not null && Activity.TurnId is { } current && current != turn) return;
            if (Activity.Phase == Phase.Failed && !start && !user && !(kind == "event_msg" && type is "user_message" or "user_input")) return;
            bool handled = true;
            if (start)
            {
                if (turn is not null && turn == Activity.TurnId && Activity.StartedAt is not null) return;
                synchronous.Clear(); asynchronous.Clear(); anonymousWait = false;
                Activity = new() { Phase = Phase.Running, Detail = "正在处理任务", StartedAt = payload.Date("started_at") ?? at, TurnId = turn, LastEventAt = at };
                return;
            }
            if (kind == "event_msg")
            {
                switch (type)
                {
                    case "task_complete": case "task_completed": case "turn_complete": case "turn_completed":
                        if (asynchronous.Count > 0 || anonymousWait) Wait(at, "等待你的回答");
                        else { Activity.Phase = Phase.Completed; Activity.Detail = "本轮执行已结束"; Activity.WaitingStartedAt = null; }
                        synchronous.Clear(); break;
                    case "turn_aborted": case "task_cancelled": case "turn_cancelled":
                        Finish(Phase.Stopped, "本轮执行已停止"); break;
                    case "task_failed": case "turn_failed":
                        Finish(Phase.Failed, "执行遇到问题，请回到 Codex 查看"); break;
                    case "request_user_input": case "user_input_requested": case "exec_approval_request": case "apply_patch_approval_request":
                        var call = payload.Text("call_id") ?? payload.Text("request_id");
                        if (call != null) synchronous.Add(call); else anonymousWait = true;
                        Wait(at, "等待你的输入或确认"); break;
                    case "user_message": case "user_input":
                        Continue(at); break;
                    case "approval_resolved":
                        var approval = payload.Text("call_id") ?? payload.Text("request_id");
                        if (approval != null && !synchronous.Remove(approval)) break;
                        anonymousWait = false;
                        if (synchronous.Count == 0 && asynchronous.Count == 0) Active(at, "已收到确认，正在继续"); break;
                    case "agent_message": case "agent_reasoning":
                        if (!Activity.Phase.IsFinished() && Activity.Phase != Phase.Waiting) Active(at, "正在处理任务"); break;
                    case "item_completed":
                        var item = payload.Field("item"); var itemType = item.Text("type")?.ToLowerInvariant();
                        if (itemType is "usermessage" or "user_message") Continue(at);
                        else if (itemType is "commandexecution" or "filechange" or "reasoning" or "mcptoolcall")
                        { if (Activity.Phase != Phase.Waiting && !Activity.Phase.IsFinished()) Active(at, "正在执行任务"); }
                        break;
                    default: handled = false; break;
                }
            }
            else if (kind == "response_item")
            {
                if (user) Continue(at);
                else if (type is "function_call" or "custom_tool_call")
                {
                    var name = payload.Text("name") ?? "";
                    if (name.EndsWith("request_user_input_async", StringComparison.Ordinal))
                    { asynchronous.Add(payload.Text("call_id") ?? "anonymous"); Wait(at, "等待你的回答"); }
                    else if (name.EndsWith("request_user_input", StringComparison.Ordinal))
                    { synchronous.Add(payload.Text("call_id") ?? "anonymous"); Wait(at, "等待你的回答"); }
                    else if (Activity.Phase != Phase.Waiting && !Activity.Phase.IsFinished()) Active(at, "正在执行任务");
                }
                else if (type is "function_call_output" or "custom_tool_call_output")
                {
                    if (payload.Text("call_id") is { } id && synchronous.Remove(id) && synchronous.Count == 0 && asynchronous.Count == 0 && !anonymousWait)
                        Active(at, "已收到回答，正在继续");
                }
                else if (type is "reasoning" or "message") { }
                else handled = false;
            }
            else handled = false;
            if (handled && at is not null) Activity.LastEventAt = at;
        }
    }
    private void Continue(DateTimeOffset? at)
    { synchronous.Clear(); asynchronous.Clear(); anonymousWait = false; Active(at, "收到输入，正在继续"); }
    private void Active(DateTimeOffset? at, string detail)
    {
        if (Activity.Phase.IsFinished() || Activity.Phase == Phase.Failed) { Activity.StartedAt = null; Activity.TurnId = null; }
        Activity.Phase = Phase.Running; Activity.Detail = detail; Activity.WaitingStartedAt = null;
    }
    private void Wait(DateTimeOffset? at, string detail)
    {
        if (Activity.Phase != Phase.Waiting)
        { if (Activity.Phase.IsFinished()) { Activity.StartedAt = null; Activity.TurnId = null; } Activity.WaitingStartedAt = at; }
        Activity.Phase = Phase.Waiting; Activity.Detail = detail;
    }
    private void Finish(Phase phase, string detail)
    { Activity.Phase = phase; Activity.Detail = detail; Activity.WaitingStartedAt = null; synchronous.Clear(); asynchronous.Clear(); anonymousWait = false; }
    public bool RecoverTiming(Activity recovered)
    {
        if (Activity.StartedAt != null || !Activity.Phase.IsActive() || Activity.Phase != recovered.Phase || Activity.LastEventAt is not { } last || last != recovered.LastEventAt ||
            recovered.StartedAt is not { } start || start > last || Activity.TurnId is { } current && current != recovered.TurnId || Activity.WaitingStartedAt is { } wait && start > wait) return false;
        Activity.StartedAt = start; Activity.TurnId ??= recovered.TurnId; return true;
    }
}
